File: app/src/main/java/com/example/Coffetable/ObjectAnalyzer.java
package com.example.Coffetable;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.graphics.Matrix;
import android.media.Image;
import android.util.Log;
import androidx.annotation.NonNull;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageProxy;
import androidx.camera.core.ExperimentalGetImage;
import org.tensorflow.lite.Interpreter;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.List;

/**
 * Analyzes images using a TensorFlow Lite model to perform object detection.
 */
public class ObjectAnalyzer implements ImageAnalysis.Analyzer {
    private static final int MODEL_INPUT_SIZE = 224; // Change according to YOLO model
    private static final int OutputSize = 10; // Set this according to your model's output

    private final Interpreter tflite;
    private final List<String> labels;

    /**
     * Constructs an ObjectAnalyzer.
     *
     * @param tflite  The TensorFlow Lite interpreter to run the model.
     * @param labels  The labels corresponding to the classes in the model.
     */
    public ObjectAnalyzer(Interpreter tflite, List<String> labels) {
        this.tflite = tflite;
        this.labels = labels;
    }

    @Override
    public void analyze(@NonNull ImageProxy image) {
        try {
            // Convert ImageProxy to Bitmap
            @SuppressLint("UnsafeOptInUsageError") Bitmap bitmap = imageToBitmap(image);
            if (bitmap == null) {
                Log.e("ObjectAnalyzer", "Image conversion to bitmap failed");
                return;
            }

            // Preprocess Bitmap for YOLO model
            ByteBuffer inputBuffer = preprocessBitmap(bitmap);

            // Run inference using TFLite
            float[][] outputBuffer = new float[1][OutputSize];
            tflite.run(inputBuffer, outputBuffer);

            // Process results
            processResults(outputBuffer);
        } catch (Exception e) {
            Log.e("ObjectAnalyzer", "Error during analysis", e);
        } finally {
            // Close the image
            image.close();
        }
    }

    /**
     * Converts an ImageProxy to a Bitmap.
     *
     * <p>This method retrieves the image from the ImageProxy and converts it from the YUV_420_888 format to
     * an RGB Bitmap format. Due to changes in the CameraX library, it uses experimental API, hence
     * the caller must opt-in to use this method.</p>
     *
     * @param image The ImageProxy containing the image data to be converted.
     * @return A Bitmap representation of the image, or null if conversion fails.
     */
    @ExperimentalGetImage // Indicates the usage of an experimental API
    private Bitmap imageToBitmap(ImageProxy image) {
        Image img = image.getImage();
        if (img == null) return null;

        // Convert YUV_420_888 to RGB Bitmap
        Bitmap bitmap = Bitmap.createBitmap(image.getWidth(), image.getHeight(), Bitmap.Config.ARGB_8888);
        YUV420888ToBitmap(img, bitmap); // You should implement this method

        // Rotate bitmap according to camera orientation
        Matrix matrix = new Matrix();
        matrix.postRotate(image.getImageInfo().getRotationDegrees());
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, true);
    }

    /**
     * Converts YUV_420_888 format image to RGB Bitmap.
     *
     * @param image The Image object representing the captured image.
     * @param bitmap The bitmap to which the image should be converted.
     */
    private void YUV420888ToBitmap(Image image, Bitmap bitmap) {
        // Implement conversion of YUV to RGB.
        // Refer to CameraX documentation or use available libraries for the conversion.
    }

    private ByteBuffer preprocessBitmap(Bitmap bitmap) {
        Bitmap resizedBitmap = Bitmap.createScaledBitmap(bitmap, MODEL_INPUT_SIZE, MODEL_INPUT_SIZE, true);

        ByteBuffer inputBuffer = ByteBuffer.allocateDirect(MODEL_INPUT_SIZE * MODEL_INPUT_SIZE * 3 * 4);
        inputBuffer.order(ByteOrder.nativeOrder());

        int[] intValues = new int[MODEL_INPUT_SIZE * MODEL_INPUT_SIZE];
        resizedBitmap.getPixels(intValues, 0, MODEL_INPUT_SIZE, 0, 0, MODEL_INPUT_SIZE, MODEL_INPUT_SIZE);

        for (int pixelValue : intValues) {
            float r = ((pixelValue >> 16) & 0xFF) / 255.0f;
            float g = ((pixelValue >> 8) & 0xFF) / 255.0f;
            float b = (pixelValue & 0xFF) / 255.0f;

            inputBuffer.putFloat(r);
            inputBuffer.putFloat(g);
            inputBuffer.putFloat(b);
        }

        return inputBuffer;
    }

    private void processResults(float[][] outputBuffer) {
        for (int i = 0; i < OutputSize; i++) {
            Log.d("ObjectAnalyzer", "Detected: " + labels.get(i) + " Confidence: " + outputBuffer[0][i]);
        }
    }
}

File: app/src/main/java/com/example/Coffetable/ObjectAnalyzer.kt
package com.example.coffetable

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.graphics.Matrix
import android.media.Image
import android.util.Log
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import androidx.camera.core.ExperimentalGetImage
import org.tensorflow.lite.Interpreter
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Analyzes images using a TensorFlow Lite model to perform object detection.
 */
class ObjectAnalyzer(
    private val tflite: Interpreter,
    private val labels: List<String>
) : ImageAnalysis.Analyzer {

    private val MODEL_INPUT_SIZE = 224 // Adjust according to your model's input size
    private val OUTPUT_SIZE = 10 // Set according to your model's output size

    override fun analyze(image: ImageProxy) {
        try {
            // Convert ImageProxy to Bitmap
            @SuppressLint("UnsafeOptInUsageError")
            val bitmap = imageToBitmap(image)
            if (bitmap == null) {
                Log.e("ObjectAnalyzer", "Image conversion to bitmap failed")
                return
            }

            // Preprocess the Bitmap for the model
            val inputBuffer: ByteBuffer = preprocessBitmap(bitmap)

            // Run inference using TensorFlow Lite
            val outputBuffer = Array(1) { FloatArray(OUTPUT_SIZE) }
            tflite.run(inputBuffer, outputBuffer)

            // Process results
            processResults(outputBuffer)
        } catch (e: Exception) {
            Log.e("ObjectAnalyzer", "Error during analysis", e)
        } finally {
            image.close() // Ensure the ImageProxy is closed
        }
    }

    @ExperimentalGetImage // Usage of an experimental API
    private fun imageToBitmap(image: ImageProxy): Bitmap? {
        val img = image.image ?: return null

        // Convert YUV_420_888 to RGB Bitmap
        val bitmap = Bitmap.createBitmap(image.width, image.height, Bitmap.Config.ARGB_8888)
        YUV420888ToBitmap(img, bitmap)

        // Rotate bitmap according to camera orientation
        val matrix = Matrix()
        matrix.postRotate(image.imageInfo.rotationDegrees.toFloat())
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
    }

    private fun YUV420888ToBitmap(image: Image, bitmap: Bitmap) {
        // Implement the conversion from YUV to RGB here.
        // This typically involves manipulating the Y, U, and V channels of the image.
    }

    private fun preprocessBitmap(bitmap: Bitmap): ByteBuffer {
        val resizedBitmap = Bitmap.createScaledBitmap(bitmap, MODEL_INPUT_SIZE, MODEL_INPUT_SIZE, true)

        val inputBuffer = ByteBuffer.allocateDirect(MODEL_INPUT_SIZE * MODEL_INPUT_SIZE * 3 * 4).apply {
            order(ByteOrder.nativeOrder())
        }

        val intValues = IntArray(MODEL_INPUT_SIZE * MODEL_INPUT_SIZE)
        resizedBitmap.getPixels(intValues, 0, MODEL_INPUT_SIZE, 0, 0, MODEL_INPUT_SIZE, MODEL_INPUT_SIZE)

        for (pixelValue in intValues) {
            val r = ((pixelValue shr 16) and 0xFF) / 255.0f
            val g = ((pixelValue shr 8) and 0xFF) / 255.0f
            val b = (pixelValue and 0xFF) / 255.0f

            inputBuffer.putFloat(r)
            inputBuffer.putFloat(g)
            inputBuffer.putFloat(b)
        }

        return inputBuffer
    }

    private fun processResults(outputBuffer: Array<FloatArray>) {
        for (i in outputBuffer[0].indices) {
            Log.d("ObjectAnalyzer", "Detected: ${labels[i]} Confidence: ${outputBuffer[0][i]}")
        }
    }
}